File: lollys-lotto-rust-sdk/src/instructions/buy_lotto_ticket.rs
use crate::instructions::*;

pub fn buy_lotto_ticket(
    round: u64,
    numbers: [u8; 6],
    authority: &Pubkey,
    user: &Pubkey,
    user_metadata: &Pubkey,
    user_usdc_token_account: &Pubkey,
    lotto_game_mint: &Pubkey,
    lotto_game: &Pubkey,
    lotto_game_vault: &Pubkey,
    lotto_ticket: &Pubkey,
    event_emitter: &Pubkey,
) -> Instruction {
    let data = lollys_lotto::instruction::BuyLottoTicket { round, numbers }.data();

    let accounts = lollys_lotto::accounts::BuyLottoTicket {
        authority: *authority,
        user: *user,
        user_metadata: *user_metadata,
        user_usdc_token_account: *user_usdc_token_account,
        lotto_game_mint: *lotto_game_mint,
        lotto_game: *lotto_game,
        lotto_game_vault: *lotto_game_vault,
        lotto_ticket: *lotto_ticket,
        event_emitter: *event_emitter,
        token_program: token::ID,
        associated_token_program: associated_token::ID,
        system_program: system_program::ID,
    }
    .to_account_metas(None);

    Instruction {
        program_id: lollys_lotto::id(),
        accounts,
        data,
    }
}

File: lollys-lotto-ts-sdk/src/codegen/instructions/buyLottoTicket.ts
import { TransactionInstruction, PublicKey, AccountMeta } from "@solana/web3.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@coral-xyz/borsh" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as types from "../types" // eslint-disable-line @typescript-eslint/no-unused-vars
import { PROGRAM_ID } from "../programId"

export interface BuyLottoTicketArgs {
  round: BN
  lottoTicketNumbers: types.LottoTicketNumbersFields
}

export interface BuyLottoTicketAccounts {
  authority: PublicKey
  user: PublicKey
  userMetadata: PublicKey
  userUsdcTokenAccount: PublicKey
  lottoGameMint: PublicKey
  lottoGame: PublicKey
  lottoGameVault: PublicKey
  lottoTicket: PublicKey
  eventEmitter: PublicKey
  tokenProgram: PublicKey
  associatedTokenProgram: PublicKey
  systemProgram: PublicKey
}

export const layout = borsh.struct([
  borsh.u64("round"),
  types.LottoTicketNumbers.layout("lottoTicketNumbers"),
])

export function buyLottoTicket(
  args: BuyLottoTicketArgs,
  accounts: BuyLottoTicketAccounts,
  programId: PublicKey = PROGRAM_ID
) {
  const keys: Array<AccountMeta> = [
    { pubkey: accounts.authority, isSigner: false, isWritable: false },
    { pubkey: accounts.user, isSigner: true, isWritable: true },
    { pubkey: accounts.userMetadata, isSigner: false, isWritable: true },
    {
      pubkey: accounts.userUsdcTokenAccount,
      isSigner: false,
      isWritable: true,
    },
    { pubkey: accounts.lottoGameMint, isSigner: false, isWritable: false },
    { pubkey: accounts.lottoGame, isSigner: false, isWritable: true },
    { pubkey: accounts.lottoGameVault, isSigner: false, isWritable: true },
    { pubkey: accounts.lottoTicket, isSigner: false, isWritable: true },
    { pubkey: accounts.eventEmitter, isSigner: false, isWritable: true },
    { pubkey: accounts.tokenProgram, isSigner: false, isWritable: false },
    {
      pubkey: accounts.associatedTokenProgram,
      isSigner: false,
      isWritable: false,
    },
    { pubkey: accounts.systemProgram, isSigner: false, isWritable: false },
  ]
  const identifier = Buffer.from([183, 179, 100, 99, 208, 96, 97, 49])
  const buffer = Buffer.alloc(1000)
  const len = layout.encode(
    {
      round: args.round,
      lottoTicketNumbers: types.LottoTicketNumbers.toEncodable(
        args.lottoTicketNumbers
      ),
    },
    buffer
  )
  const data = Buffer.concat([identifier, buffer]).subarray(0, 8 + len)
  const ix = new TransactionInstruction({ keys, programId, data })
  return ix
}
